File: binding/src/libqgesturerecognizer.cpp
#include <QEvent>
#include <QGesture>
#include <QGestureRecognizer>
#include <QObject>
#include <qgesturerecognizer.h>
#include "libqgesturerecognizer.h"
#include "libqgesturerecognizer.hxx"

QGestureRecognizer* QGestureRecognizer_new() {
	 return new VirtualQGestureRecognizer();
}

QGestureRecognizer* QGestureRecognizer_new2(const QGestureRecognizer* param1) {
	 return new VirtualQGestureRecognizer(*param1);
}

QGesture* QGestureRecognizer_Create(QGestureRecognizer* self, QObject* target) {
	return self->create(target);
}

int QGestureRecognizer_Recognize(QGestureRecognizer* self, QGesture* state, QObject* watched, QEvent* event) {
	return self->recognize(state, watched, event);
}

void QGestureRecognizer_Reset(QGestureRecognizer* self, QGesture* state) {
	self->reset(state);
}

int QGestureRecognizer_RegisterRecognizer(QGestureRecognizer* recognizer) {
	return QGestureRecognizer::registerRecognizer(recognizer);
}

void QGestureRecognizer_UnregisterRecognizer(int typeVal) {
	QGestureRecognizer::unregisterRecognizer(static_cast<Qt::GestureType>(typeVal));
}

void QGestureRecognizer_OperatorAssign(QGestureRecognizer* self, const QGestureRecognizer* param1) {
	self->operator=(*param1);
}

// Base class handler implementation
QGesture* QGestureRecognizer_QBaseCreate(QGestureRecognizer* self, QObject* target) {
	auto* vqgesturerecognizer = dynamic_cast<VirtualQGestureRecognizer*>(self);
	if (vqgesturerecognizer && vqgesturerecognizer->isVirtualQGestureRecognizer) {
vqgesturerecognizer->setQGestureRecognizer_Create_IsBase(true);
	return vqgesturerecognizer->create(target);
}
}

// Auxiliary method to allow providing re-implementation
void QGestureRecognizer_OnCreate(QGestureRecognizer* self, intptr_t slot) {
	auto* vqgesturerecognizer = dynamic_cast<VirtualQGestureRecognizer*>(self);
	if (vqgesturerecognizer && vqgesturerecognizer->isVirtualQGestureRecognizer) {
vqgesturerecognizer->setQGestureRecognizer_Create_Callback(reinterpret_cast<VirtualQGestureRecognizer::QGestureRecognizer_Create_Callback>(slot));
}
}

// Base class handler implementation
int QGestureRecognizer_QBaseRecognize(QGestureRecognizer* self, QGesture* state, QObject* watched, QEvent* event) {
	auto* vqgesturerecognizer = dynamic_cast<VirtualQGestureRecognizer*>(self);
	if (vqgesturerecognizer && vqgesturerecognizer->isVirtualQGestureRecognizer) {
vqgesturerecognizer->setQGestureRecognizer_Recognize_IsBase(true);
	return vqgesturerecognizer->recognize(state, watched, event);
}
}

// Auxiliary method to allow providing re-implementation
void QGestureRecognizer_OnRecognize(QGestureRecognizer* self, intptr_t slot) {
	auto* vqgesturerecognizer = dynamic_cast<VirtualQGestureRecognizer*>(self);
	if (vqgesturerecognizer && vqgesturerecognizer->isVirtualQGestureRecognizer) {
vqgesturerecognizer->setQGestureRecognizer_Recognize_Callback(reinterpret_cast<VirtualQGestureRecognizer::QGestureRecognizer_Recognize_Callback>(slot));
}
}

// Base class handler implementation
void QGestureRecognizer_QBaseReset(QGestureRecognizer* self, QGesture* state) {
	auto* vqgesturerecognizer = dynamic_cast<VirtualQGestureRecognizer*>(self);
	if (vqgesturerecognizer && vqgesturerecognizer->isVirtualQGestureRecognizer) {
vqgesturerecognizer->setQGestureRecognizer_Reset_IsBase(true);
	vqgesturerecognizer->reset(state);
}
}

// Auxiliary method to allow providing re-implementation
void QGestureRecognizer_OnReset(QGestureRecognizer* self, intptr_t slot) {
	auto* vqgesturerecognizer = dynamic_cast<VirtualQGestureRecognizer*>(self);
	if (vqgesturerecognizer && vqgesturerecognizer->isVirtualQGestureRecognizer) {
vqgesturerecognizer->setQGestureRecognizer_Reset_Callback(reinterpret_cast<VirtualQGestureRecognizer::QGestureRecognizer_Reset_Callback>(slot));
}
}

void QGestureRecognizer_Delete(QGestureRecognizer* self) {
    delete self;
}

File: binding/include/libqgesturerecognizer.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qgesturerecognizer.html
pub const qgesturerecognizer = struct {

    /// New constructs a new QGestureRecognizer object.
    pub fn New() QtC.QGestureRecognizer {
        return qtc.QGestureRecognizer_new();
    }


    /// New2 constructs a new QGestureRecognizer object.
    pub fn New2(param1: ?*anyopaque) QtC.QGestureRecognizer {
        return qtc.QGestureRecognizer_new2(@ptrCast(param1));
    }


    pub fn Create(self: ?*anyopaque, target: ?*anyopaque) QtC.QGesture {
        return qtc.QGestureRecognizer_Create(@ptrCast(self), @ptrCast(target));
    }

    /// Allows for overriding the related default method
    pub fn OnCreate(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque) callconv(.c) QtC.QGesture) void {
        qtc.QGestureRecognizer_OnCreate(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseCreate(self: ?*anyopaque, target: ?*anyopaque) QtC.QGesture {
        return qtc.QGestureRecognizer_QBaseCreate(@ptrCast(self), @ptrCast(target));
    }

    pub fn Recognize(self: ?*anyopaque, state: ?*anyopaque, watched: ?*anyopaque, event: ?*anyopaque) i32 {
        return qtc.QGestureRecognizer_Recognize(@ptrCast(self), @ptrCast(state), @ptrCast(watched), @ptrCast(event));
    }

    /// Allows for overriding the related default method
    pub fn OnRecognize(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque, ?*anyopaque, ?*anyopaque) callconv(.c) i32) void {
        qtc.QGestureRecognizer_OnRecognize(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseRecognize(self: ?*anyopaque, state: ?*anyopaque, watched: ?*anyopaque, event: ?*anyopaque) i32 {
        return qtc.QGestureRecognizer_QBaseRecognize(@ptrCast(self), @ptrCast(state), @ptrCast(watched), @ptrCast(event));
    }

    pub fn Reset(self: ?*anyopaque, state: ?*anyopaque) void {
        qtc.QGestureRecognizer_Reset(@ptrCast(self), @ptrCast(state));
    }

    /// Allows for overriding the related default method
    pub fn OnReset(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque) callconv(.c) void) void {
        qtc.QGestureRecognizer_OnReset(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseReset(self: ?*anyopaque, state: ?*anyopaque) void {
        qtc.QGestureRecognizer_QBaseReset(@ptrCast(self), @ptrCast(state));
    }

    pub fn RegisterRecognizer(recognizer: ?*anyopaque) i64 {
        return qtc.QGestureRecognizer_RegisterRecognizer(@ptrCast(recognizer));
    }

    pub fn UnregisterRecognizer(typeVal: i64) void {
        qtc.QGestureRecognizer_UnregisterRecognizer(@intCast(typeVal));
    }

    pub fn OperatorAssign(self: ?*anyopaque, param1: ?*anyopaque) void {
        qtc.QGestureRecognizer_OperatorAssign(@ptrCast(self), @ptrCast(param1));
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QGestureRecognizer_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/gesturerecognizer.html#types
pub const enums = struct {
    pub const ResultFlag = enum {
        pub const Ignore: i32 = 1;
        pub const MayBeGesture: i32 = 2;
        pub const TriggerGesture: i32 = 4;
        pub const FinishGesture: i32 = 8;
        pub const CancelGesture: i32 = 16;
        pub const ResultState_Mask: i32 = 255;
        pub const ConsumeEventHint: i32 = 256;
        pub const ResultHint_Mask: i32 = 65280;
    };

};
